azure-mgmt-compute/src/test/java/com/microsoft/azure/management/compute/VirtualMachineExtensionOperationsTests.java

```diff
@@ -0,0 +1,119 @@
+                .withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_14_04_LTS)
+                .withRootUserName("Foo12")
+                .withPassword("BaR@12abc!")
+                .withSize(VirtualMachineSizeTypes.STANDARD_D3)
+                .create();
+
+        // Using VMAccess Linux extension to reset the password for the existing user 'Foo12'
+        // https://github.com/Azure/azure-linux-extensions/blob/master/VMAccess/README.md
+        //
+        vm.update()
+                .defineNewExtension("VMAccessForLinux")
+                    .withPublisher("Microsoft.OSTCExtensions")
+                    .withType("VMAccessForLinux")
+                    .withVersion("1.4")
+                    .withProtectedSetting("username", "Foo12")
+                    .withProtectedSetting("password", "B12a6@12xyz!")
+                    .withProtectedSetting("reset_ssh", "true")
+                .attach()
+                .apply();
+
+        Assert.assertTrue(vm.extensions().size() > 0);
+        Assert.assertTrue(vm.extensions().containsKey("VMAccessForLinux"));
+
+        // Update the VMAccess Linux extension to reset password again for the user 'Foo12'
+        //
+        vm.update()
+                .updateExtension("VMAccessForLinux")
+                    .withProtectedSetting("username", "Foo12")
+                    .withProtectedSetting("password", "muy!234OR")
+                    .withProtectedSetting("reset_ssh", "true")
+                .parent()
+                .apply();
+    }
+
+    @Test
+    public void canInstallUninstallCustomExtension() throws Exception {
+        final String RG_NAME = ResourceNamer.randomResourceName("vmexttest", 15);
+        final String LOCATION = "eastus";
+        final String VMNAME = "javavm";
+
+        final String mySqlInstallScript = "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/4397e808d07df60ff3cdfd1ae40999f0130eb1b3/mysql-standalone-server-ubuntu/scripts/install_mysql_server_5.6.sh";
+        final String installCommand = "bash install_mysql_server_5.6.sh Abc.123x(";
+        List<String> fileUris = new ArrayList<>();
+        fileUris.add(mySqlInstallScript);
+
+        // Create Linux VM with a custom extension to install MySQL
+        //
+        VirtualMachine vm = computeManager.virtualMachines()
+                .define(VMNAME)
+                .withRegion(LOCATION)
+                .withNewResourceGroup(RG_NAME)
+                .withNewPrimaryNetwork("10.0.0.0/28")
+                .withPrimaryPrivateIpAddressDynamic()
+                .withoutPrimaryPublicIpAddress()
+                .withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_14_04_LTS)
+                .withRootUserName("Foo12")
+                .withPassword("BaR@12abc!")
+                .withSize(VirtualMachineSizeTypes.STANDARD_D3)
+                .defineNewExtension("CustomScriptForLinux")
+                    .withPublisher("Microsoft.OSTCExtensions")
```

Comment:
Interms of choosing the extension image, user can either do:

```java
.defineNewExtension(..)
  .withPublisher(..)
  .withType(..)
  .withVersion(..)
  .withPublicSettings(..)
  .withPrivateSettings(..)
  .....
```

OR

```java
.defineNewExtension(..)
  .withImage(VirtualMachineExtensionImage image)
  .withPublicSettings(..)
  .withPrivateSettings(..)
 .....
```

and the `image` can be fetched by

```java
List<VirtualMachineExtensionImage> extensions = ComputeManager().extensions().listByRegion(..);
```

I thought about providing enum for some popular extensions such as CustomScript / VMAccess. That can be done, but it still user has to find out the public and private settings for the extension he want to use. These settings are `Object` (like template and it's parameters) not any specialized type. If we want to make it more valuable then we could directly expose definition + withers for the same in VM define/update flow.

```
   virtualMachines()
     .define()
     ...
     .withCustomScriptExtension()
        .withScriptUri(..)
        .withScriptUri(..) 
        .withCommandToExecute(..)
        .attach()
```

Regarding VMAccess, it is applicable for only Linux and extensions are optionals, so when we are in optional flow we don't whether or not to show this depending on OS.